src/lib/shared/form/form.stories.tsx

```diff
@@ -0,0 +1,71 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import React, { useCallback } from 'react'
+import { ComponentStory, ComponentMeta } from '@storybook/react'
+import { FormProvider, useFieldProps, useForm } from './form'
+import { Values, FieldValidatorFn } from './types'
+type MyFormValues = { foo: string; bar: Date }
+const MyForm = (_: {}) => {
+  const onSubmit = useCallback((values: MyFormValues) => {
+    return new Promise<MyFormValues>((resolve) => {
+      setTimeout(() => resolve(values), 1000)
+    })
+  }, [])
+  const form = useForm<MyFormValues>({ onSubmit, initialValues: { foo: '', bar: new Date() } })
+  return (
+    <FormProvider {...form}>
+      <form onSubmit={form.submit}>
+        <MyField<MyFormValues> name="foo" />
+        <MyDateField<MyFormValues> name="bar" />
+        <input disabled={!form.isValid || form.isSubmitting} type="submit" value="Click me!" />
+      </form>
+    </FormProvider>
+  )
+}
+
+const MyField = <T extends Values>({ name }: { name: keyof T }) => {
+  const validate: FieldValidatorFn<T, keyof T> = useCallback(
+    (value, _) =>
+      new Promise((resolve) =>
+        setTimeout(
+          () => resolve(!value ? 'Required' : (value as unknown as string).length < 3 ? 'Too short' : undefined),
+          300
+        )
+      ),
+    []
+  )
+  const props = useFieldProps({ name, validate, type: 'text' })
+  return <input {...props} />
+}
+
+const MyDateField = <T extends Values>({ name }: { name: keyof T }) => {
+  const validate: FieldValidatorFn<T, keyof T> = useCallback(
+    (value, _) =>
+      new Promise((resolve) =>
+        setTimeout(() => resolve((value as Date).getMilliseconds() < 1000000 ? 'Invalid' : undefined), 300)
+      ),
+    []
+  )
+  const props = useFieldProps<T>({
+    name,
+    validate,
+    type: 'date',
+    fromString: (v) => new Date(v) as any,
+    asString: (v: any) => v?.toISOString(),
+  })
+  return <input {...props} />
+}
+
+// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
+export default {
+  title: 'Lib/Form',
```

Comment:
```suggestion
  title: 'Components/Form',
```

I would also suggest to put the form into the components directory. Although I am not quite sure what the purpose of the "components" and "shared" directories is...

Reply:
This is not a component though, it's an library to be used in conjunction with other components

Reply:
Hm, but if it's not a component, do we need a story for that?

Reply:
Story is imho essentially a usage example, so even if the code does not add markup, or visuals, but adds behaviour instead (that's eseentially what's happening here), then I'd say it's even more important to have this kind of interactive sandbox that a story provides.

Reply:
Ok I understand!